src/ip.rs

```diff
@@ -1,4 +1,283 @@
-use std::net::{IpAddr, Ipv4Addr, Ipv6Addr};
+use std::{
+    fmt,
+    net::{AddrParseError, IpAddr, Ipv4Addr, Ipv6Addr},
+    num::ParseIntError,
+    str::FromStr,
+};
+
+/// An address withing a subnet.
+pub(crate) struct ScopedIpAddr {
+    pub(crate) addr: IpAddr,
+    #[allow(dead_code)]
+    pub(crate) subnet: IpSubnet,
```

Comment:
I'm struggling a bit with the subnet part. This factoring makes it feel like the host _has a subnet_, since the host has one or more of these.

I wonder if we should get the network interfaces right first and then add subnet support.